### PR TITLE
Add FollowAction to allow retrying current step

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#711

# Overview

- [x] I have added a `FollowAction` to the return-type of the `follow`-callback. It allows the callback to tell the `follow` function to retry the current step (which is needed for stake-pools).

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -520,14 +525,21 @@ restoreWallet
     -> ExceptT ErrNoSuchWallet IO ()
 restoreWallet ctx wid = do
     cps <- liftIO $ DB.listCheckpoints db (PrimaryKey wid)
-    let forward = restoreBlocks @ctx @s @t @k ctx wid
-    let backward = DB.rollbackTo db (PrimaryKey wid)
-    liftIO $ follow nw tr cps forward backward (view #header)
+    let forward bs h = run $ restoreBlocks @ctx @s @t @k ctx wid bs h
+    let backward sid = run $ DB.rollbackTo db (PrimaryKey wid) sid
+    _ <- liftIO $ follow nw tr cps forward backward (view #header)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: small thing, but prefer `void $ ...` to `_ <-`

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 2]
```diff
@@ -233,19 +234,31 @@ instance Functor (NextBlocksResult target) where
...
     -> Trace IO Text
     -- ^ Logger trace
     -> [BlockHeader]
     -- ^ A list of known tips to start from. Blocks /after/ the tip will be yielded.
-    -> (NE.NonEmpty block -> BlockHeader -> ExceptT e0 IO ())
+    -> (NE.NonEmpty block -> BlockHeader -> IO (FollowAction e))
     -- ^ Callback with blocks and the current tip of the /node/.
     -- @follow@ stops polling and terminates if the callback errors.
-    -> (SlotId -> ExceptT e1 IO ())
+    -> (SlotId -> IO (FollowAction e))
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs]: These were `e0` and `e1` on purpose actually. The errors returned by these two callbacks might be different and we can't just force them to be of the same type.

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 3]
```diff
@@ -296,19 +309,20 @@ follow nl tr cps yield rollback header =
...
+                >>= handle cursor' "Failed to roll backward: "
+      where
+        handle cursor' msg x = case x of
+                ExitWith e ->
+                    logError tr $ T.pack $
+                        msg <> show e
+                Continue ->
                     step cursor'
+                RetryCurrentStep ->
+                    step cursor
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs:351]: :+1:

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 4]
```diff
@@ -233,19 +234,31 @@ instance Functor (NextBlocksResult target) where
...
+-- | @FollowAction@ enables the callback of @follow@ to signal if the
+-- chain-following should @Quit@, @Continue@, or of the current callback should
+-- be forgotten and retried (@RetryCurrentStep@).
+--
+-- NOTE: @RetryCurrentStep@ is needed to handle data-races in
+-- 'Cardano.Pool.Metrics'. Aside for that it is arguably a bit out-of-place.
+data FollowAction err
+    = ExitWith err
+    | Continue
+    | RetryCurrentStep
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs]: What about `Retry`

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 5]
```diff
@@ -233,19 +234,31 @@ instance Functor (NextBlocksResult target) where
         RollForward cur bh bs -> RollForward cur bh (fmap f bs)
         RollBackward cur -> RollBackward cur
 
+-- | @FollowAction@ enables the callback of @follow@ to signal if the
+-- chain-following should @Quit@, @Continue@, or of the current callback should
+-- be forgotten and retried (@RetryCurrentStep@).
+--
+-- NOTE: @RetryCurrentStep@ is needed to handle data-races in
+-- 'Cardano.Pool.Metrics'. Aside for that it is arguably a bit out-of-place.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs]: That last sentence is a bit off :upside_down_face: 

```
-- RetryCurrentStep gives some control to chain followers who might have failed to roll forward but do not require the chain follower to exit. 
```

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 6]
```diff
@@ -233,19 +234,31 @@ instance Functor (NextBlocksResult target) where
         RollForward cur bh bs -> RollForward cur bh (fmap f bs)
         RollBackward cur -> RollBackward cur
 
+-- | @FollowAction@ enables the callback of @follow@ to signal if the
+-- chain-following should @Quit@, @Continue@, or of the current callback should
+-- be forgotten and retried (@RetryCurrentStep@).
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs]: Comment need to be updated with the new constructors' names :+1:

File: lib/core/src/Cardano/Wallet/Network.hs (hunk 4 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Network.hs]: Sure.

File: lib/core/src/Cardano/Wallet/Network.hs (hunk 2 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Network.hs]: 🤔 Sure. I don't think that there is an actual reason in practice though.

Btw, on the side I'm kind of wondering wouldn't be nicer to combine `rollback` and `yield` into one function.

**Review by KtorZ — APPROVED** [review]

**Pull request merged**